### PR TITLE
feat: support process.getBuiltinModule

### PR DESCRIPTION
**Pull request by watson**
A new way to require built-in Node.js modules (a.k.a. core modules) was added in Node.js v22.3.0 (PR: nodejs/node#52762):

```js
require('http') === process.getBuiltinModule('http')
```

The way this module patches the `require` function means we don't catch calls to `process.getBuiltinModule`, which then remain unpatched. This PR fixes that by also patching calls to `process.getBuiltinModule` directly.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -151,6 +170,11 @@ function Hook (modules, options, onrequire) {
           filename = idWithoutPrefix
         }
       }
+    } else if (coreOnly) {
+      // `coreOnly` is `true` if this was a call to `process.getBuiltinModule`, in which case
+      // we don't want to return anything if the requested `id` isn't a core module
+      debug('call to process.getBuiltinModule with unknown built-in id - returning undefined')
+      return undefined
```
Comment (trentm) [inline index.js:177]: Or perhaps call the original to be sure to get its behaviour. I realize it is returning undefined, and isn't likely to change. Calling the original could handle the same case as in the `} else {` block just below where there was double-monkey-patching of `require`.

```suggestion
      return self._origGetBuiltinModule.apply(this, args)
```

File: index.js [hunk 2]
```diff
@@ -139,6 +140,24 @@ function Hook (modules, options, onrequire) {
       return self._origRequire.apply(this, arguments)
     }
 
+    return patchedRequire.call(this, arguments, false)
+  }
+
+  this._getBuiltinModule = process.getBuiltinModule = function (id) {
```
Comment (trentm) [inline index.js:146]: I don't think we should be setting `process.getBuiltinModule` in versions of Node.js that don't have it.
Thoughts?

**Review by trentm — CHANGES_REQUESTED** [review]
Nice. s/require-in-the-middle/require-and-get-builtin-module-in-the-middle/.

**Comment by trentm** [comment]
Before I forget, I think the README should also get mention that this new core API would get patched as well.

File: index.js (hunk 2 above)
Comment (watson) [reply index.js]: Fixed in eb50f6d 👍

File: index.js (hunk 1 above)
Comment (watson) [reply index.js]: Fixed in eb50f6d 👍

**Comment by watson** [comment]
@trentm Thanks for the review. I've addressed your comments in eb50f6d

File: index.js [hunk 3]
```diff
@@ -288,11 +317,19 @@ function Hook (modules, options, onrequire) {
...
+
   if (this._require === Module.prototype.require) {
     Module.prototype.require = this._origRequire
-    debug('unhook successful')
+    debug('require unhook successful')
+  } else {
+    debug('require unhook unsuccessful')
+  }
+
+  if (this._getBuiltinModule === process.getBuiltinModule) {
```
Comment (trentm) [inline index.js:328]: Can we also guard this on `typeof process.getBuiltinModule === 'function'` or `!== undefined`? Currently `undefined === undefined` and the debug logging is misleading on earlier versions of Node.js:

```
% cat foo.js
console.log('Running node', process.version)
var {Hook} = require('./');
var hook = new Hook(['http'], (mod, name, baseDir) => {
    console.log('Hooked %j', name);
});
require('http');
hook.unhook();

% DEBUG=require-in-the-middle node foo.js
Running node v18.20.4
  require-in-the-middle registering require hook +0ms
  require-in-the-middle processing core module require('http'): http +0ms
  require-in-the-middle calling require hook: http +1ms
Hooked "http"
  require-in-the-middle returning module: http +0ms
  require-in-the-middle require unhook successful +0ms
  require-in-the-middle process.getBuiltinModule unhook successful +0ms
```

**Review by trentm — CHANGES_REQUESTED** [review]

File: index.js (hunk 3 above)
Comment (watson) [reply index.js]: Good catch! Thanks for such a thorough review 💯 

Fixed now in 93cf4ec 👍

**Review by trentm — APPROVED** [review]

**Comment by trentm** [comment]
Thanks, @watson!  I'll get a release out soonish.

**Pull request merged**